backend/server/src/main.rs

```diff
@@ -79,8 +83,30 @@ async fn main() -> anyhow::Result<()> {
                 .to_string())
     );
 
-    tokio::spawn(async { unix_service.run().await });
-    web_service.run().await;
+    let cancellation_token = CancellationToken::new();
+
+    let eval_handle = tokio::spawn(eval_service.run(cancellation_token.clone()));
+    let unix_handle = tokio::spawn(unix_service.run(cancellation_token.clone()));
+    let web_handle = tokio::spawn(web_service.run(cancellation_token.clone()));
+
+    let mut sigterm = signal(SignalKind::terminate()).context("failed to get sigterm handle")?;
+    let mut sigint = signal(SignalKind::interrupt()).context("failed to get sigint handle")?;
+
+    tokio::select! {
+        _ = sigterm.recv() => {
+            info!("Received SIGTERM, gracefully shutting down");
+        }
+        _ = sigint.recv() => {
+            info!("Received SIGINT, gracefully shutting down");
+        }
+    }
+
+    cancellation_token.cancel();
+
+    // Wait for the services to shutdown
+    _ = tokio::join!(eval_handle, unix_handle, web_handle);
```

Comment:
Similarly to above, if you want to avoid the macro call, you can spawn these tasks in a `JoinSet` and then use `JoinSet::join_all`. I am however fine with leaving this as is.

Reply:
it's fine as is